### PR TITLE
Fewer logging of non-leaders in debug log-level

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Fixes: #351

### CODE REVIEW

**Review by cevian — CHANGES_REQUESTED** [review]
For debug printing I think it would be better to rate-limit these messages (say print every 1 minute) rather than just once. The reason is in debug mode you want logging to be verbose.

File: pkg/log/log.go [hunk 1]
```diff
@@ -108,3 +112,38 @@ func parseLogLevel(logLevel string) (level.Option, error) {
...
+func timedLogger() {
+	t := time.NewTicker(logOnceTimedDuration)
+	for range t.C {
+		logStoreMux.Lock()
+		for k, logMsg := range logStore {
+			Debug("msg", "timed-debug")
+			Debug(logMsg...)
+			delete(logStore, k)
+		}
+		if len(logStore) == 0 {
```
Comment (cevian) [inline pkg/log/log.go:129]: This will always, be true no?

File: pkg/log/log.go [hunk 2]
```diff
@@ -108,3 +112,38 @@ func parseLogLevel(logLevel string) (level.Option, error) {
...
+
+// timedLogger logs from logStore every logOnceTimedDuration. It deletes the log entry from the store
+// after it has logged once.
+func timedLogger() {
+	t := time.NewTicker(logOnceTimedDuration)
+	for range t.C {
+		logStoreMux.Lock()
+		for k, logMsg := range logStore {
+			Debug("msg", "timed-debug")
+			Debug(logMsg...)
```
Comment (cevian) [inline pkg/log/log.go:126]:
```suggestion
			Debug("kind", "rate-limited", logMsg...)
```

File: pkg/log/log.go (hunk 1 above)
Comment (cevian) [reply pkg/log/log.go:129]: in fact I'm ok with a panic if this isn't true

File: pkg/log/log.go [hunk 3]
```diff
@@ -108,3 +112,38 @@ func parseLogLevel(logLevel string) (level.Option, error) {
...
+		if len(logStore) == 0 {
+			logStoreMux.Unlock()
+			return
+		}
+		logStoreMux.Unlock()
+	}
+}
+
+// DebugTimed logs Debug level logs once in every logOnceTimedDuration.
+func DebugTimed(keyvals ...interface{}) {
```
Comment (cevian) [inline pkg/log/log.go:138]:
```suggestion
func DebugRateLimited(keyvals ...interface{}) {
```

File: pkg/log/log.go [hunk 4]
```diff
@@ -108,3 +112,38 @@ func parseLogLevel(logLevel string) (level.Option, error) {
 		return nil, fmt.Errorf("unrecognized log level %q", logLevel)
 	}
 }
+
+const logOnceTimedDuration = time.Second * 10
+
+// timedLogger logs from logStore every logOnceTimedDuration. It deletes the log entry from the store
+// after it has logged once.
+func timedLogger() {
+	t := time.NewTicker(logOnceTimedDuration)
```
Comment (cevian) [inline pkg/log/log.go:121]: I think this only needs to fire once, so not a ticker?

**Review by cevian — CHANGES_REQUESTED** [review]

File: pkg/log/log.go (hunk 1 above)
Comment (Harkishen-Singh) [reply pkg/log/log.go:129]: Yeah, this should always be true. Here, just for sanitary checks.

File: pkg/log/log.go (hunk 2 above)
Comment (Harkishen-Singh) [reply pkg/log/log.go:126]: That will lead to compilation errors. However, I will implement a work-around here.

File: pkg/log/log.go [hunk 5]
```diff
@@ -108,3 +112,42 @@ func parseLogLevel(logLevel string) (level.Option, error) {
...
+	for k, logMsg := range logStore {
+		Debug(applyKind(logMsg)...)
+		delete(logStore, k)
+	}
+	if len(logStore) != 0 {
+		panic(fmt.Sprintf("logStore not empty after flushing rate-limited logs: expected size to be 0 but received %d", len(logStore)))
+	}
+	// delete() does not remove the value from the map. It just sets the flag, meaning memory will increase even though less (at first).
+	// Hence, better to re-initialize it within the lock.
+	logStore = make(map[string][]interface{})
```
Comment (cevian) [inline pkg/log/log.go:138]: if we do this then we don't need the delete above (delete can get expensive), Or the panic. I like that more.

File: pkg/log/log.go [hunk 6]
```diff
@@ -108,3 +112,42 @@ func parseLogLevel(logLevel string) (level.Option, error) {
...
+
+// timedLogger logs from logStore every logOnceTimedDuration. It deletes the log entry from the store
+// after it has logged once.
+func timedLogger() {
+	applyKind := func(logMsg []interface{}) (newLogMsg []interface{}) {
+		newLogMsg = append(newLogMsg, "kind", "rate-limited", "duration", logOnceTimedDuration.String())
+		newLogMsg = append(newLogMsg, logMsg...)
+		return
+	}
+	time.Sleep(logOnceTimedDuration)
```
Comment (cevian) [inline pkg/log/log.go:126]: nit: let's make this the first line in the function

File: pkg/log/log.go (hunk 5 above)
Comment (Harkishen-Singh) [reply pkg/log/log.go:138]: >  I like that more

If we just do the delete, then we aren't clearing the memory used by the map for storing log. That's the behaviour of delete() in map. Hence, I reinitialized it.

Comment (Harkishen-Singh) [reply pkg/log/log.go:138]: I think, going with just initialization should be better way here in terms of efficiency and memory. Do you agree?

Comment (cevian) [reply pkg/log/log.go:132]: sorry I meant that I like the initialization more - we should remove the delete and the panic

**Review by cevian — APPROVED** [review]

File: pkg/log/log.go (hunk 5 above)
Comment (Harkishen-Singh) [reply pkg/log/log.go:132]: oh, got it. The current was in line with that method. No prob.

**Pull request merged**